docs/src/main/tut/typeclasses/lawtesting.md

```diff
@@ -0,0 +1,133 @@
+
+```tut:book
+import cats.tests.CatsSuite
+
+class TreeLawTests extends CatsSuite {
+
+}
+```
+
+The key to testing laws is the `checkAll` function, which takes a name for your test and a Discipline ruleset.
+Cats has defined rulesets for all type class laws in `cats.laws.discipline.*`.
+
+So for our example we will want to import `cats.laws.discipline.FunctorTests` and call `checkAll` with it.
+Before we do so, however,
+we will have to bring our instances into scope as well as the derived `Arbitrary` instances from `scalacheck-shapeless`:
+
+
+```tut:book
+import Tree._
+
+import org.scalacheck.Shapeless._
+
+import cats.laws.discipline.FunctorTests
+
+class TreeLawTests extends CatsSuite {
+  checkAll("Tree.FunctorLaws", FunctorTests[Tree].functor[Int, Int, String])
+}
+```
+
+Now when we run `test` in our sbt console, ScalaCheck will test if the `Functor` laws hold for our `Tree` type.
+You should see something like this:
+
+```
+[info] TreeLawTests:
+[info] - Tree.FunctorLaws.functor.covariant composition
+[info] - Tree.FunctorLaws.functor.covariant identity
+[info] - Tree.FunctorLaws.functor.invariant composition
+[info] - Tree.FunctorLaws.functor.invariant identity
+[info] ScalaTest
+[info] Run completed in 537 milliseconds.
+[info] Total number of tests run: 4
+[info] Suites: completed 1, aborted 0
+[info] Tests: succeeded 4, failed 0, canceled 0, ignored 0, pending 0
+[info] All tests passed.
+[info] Passed: Total 4, Failed 0, Errors 0, Passed 4
+[success] Total time: 1 s, completed Aug 31, 2017 2:19:22 PM
+```
+
+And voila, you've successfully proven that your data type upholds the Functor laws!
+
+### Testing cats.kernel instances
+
+For most of the type classes included in cats, the above will work great.
+However, the law tests for the type classes inside `cats.kernel` are structured a bit differently.
+These include `Semigroup`, `Monoid`, `Group` and `Semilattice`.
+Instead of importing the laws from `cats.laws.discipline.*`, we have to import `cats.kernel.laws.GroupLaws`
+and then call the corresponding method, e.g. `GroupLaws[Foo].monoid`, or `GroupLaws[Foo].semigroup`.
+
+Let's test it out, by defining a `Semigroup` instance for our `Tree` type.
+
+```tut:book
+import cats.implicits._
+
+implicit def semigroupTree[A: Semigroup]: Semigroup[Tree[A]] = new Semigroup[Tree[A]] {
+  def combine(x: Tree[A], y: Tree[A]) = (x, y) match {
+    case (Leaf, _) => Leaf
+    case (_, Leaf) => Leaf
+    case (Node(xp, xLeft, xRight), Node(yp, yLeft, yRight)) =>
+      Node(xp |+| yp, xLeft |+| yLeft, xRight |+| yRight)
+  }
+}
+```
+
+Then we can again test the instance inside our class extending `CatsSuite`:
+
+```tut:book
+import cats.laws.discipline.FunctorTests
+import cats.kernel.laws.GroupLaws
+
+class TreeLawTests extends CatsSuite {
+  checkAll("Tree[Int].MonoidLaws", GroupLaws[Tree[Int]].semigroup)
+  checkAll("Tree.FunctorLaws", FunctorTests[Tree].functor[Int, Int, String])
+}
+```
```

Comment:
Note that with catalysts you can do something similar to how algebra/kernel calls the laws, which is something like:

```scala
...
import catalysts.macros.TypeTagM

// this can be added to the general test suite
implicit def groupLaws[A: Eq: Arbitrary]: GroupLaws[A] = GroupLaws[A]

....
laws[GroupLaws, Tree[Int]].check(_.semigroup)
```

Reply:
Do you think this is worth mentioning? I'm not really sure if I can see the benefit 🤔